libipl/libipl.H

```diff
@@ -18,11 +21,50 @@ enum ipl_type {
 #define IPL_INFO    1
 #define IPL_DEBUG   2
 
+//IPL Error types
+enum ipl_error_type {
+	IPL_NO_ERR = 0,
+	IPL_HWP_ERR,
+	IPL_SBE_CHIPOP_ERR,
+	IPL_SBE_BOOT_FAIL,
+	IPL_PRI_PROC_NON_FUNC,
+	IPL_FSI_TGT_NOT_FOUND,
+	IPL_CFAM_ERR,
+};
+
+typedef std::map<ipl_error_type, const char*> err_msg_map_type;
+const err_msg_map_type err_msg_map = {
+	{IPL_NO_ERR, ""},
+	{IPL_HWP_ERR, "HWP execution reported error"},
+	{IPL_SBE_CHIPOP_ERR, "SBE chip-op reported error"},
+	{IPL_SBE_BOOT_FAIL, "SBE Boot failure"},
+	{IPL_PRI_PROC_NON_FUNC, "Primary processor is non functional"},
+	{IPL_FSI_TGT_NOT_FOUND, "FSI target is not available"},
+	{IPL_CFAM_ERR, "CFAM read/write failure"},
+};
+
+// Error info structure.
+struct ipl_error_info {
+	enum ipl_error_type type;
+	//additional debug/callout data need to be included for cfam failures
+	void *private_data;
+	//specilized constructors
+	ipl_error_info() : type(IPL_NO_ERR), private_data(NULL)  {}
+	ipl_error_info(ipl_error_type type) : type(type), private_data(NULL) {}
+
+	~ipl_error_info(){
+		// freeup the private_data allocated memory
+		if (private_data){
+			free(private_data);
+		}
+	}
+};
+
```

Comment:
Does this have to be in header file?

Just for my information, since when can structures have constructors and destructors?  Or C++ treats structures also like classes?

Reply:
>Does this have to be in header file?
readability point i kept destructor also here.

"The C++ class is an extension of the C language structure. Because the only difference between a structure and a class is that structure members have public access by default and class members have private access by default, you can use the keywords class or struct to define equivalent classes."

Reply:
as the structure is used in functions declared in header files it needs to be in header files. (scoping) or else compiler will say underfined variable.